File: src-tauri/src/commands/telemetry.rs
//! Frontend entry points for usage counters. Rust-side actions record
//! themselves directly via `crate::telemetry`; these commands exist for
//! interactions that only the UI sees (navigation clicks, view toggles,
//! export choices). Names are validated in `crate::telemetry::valid_key`.
use crate::state::AppState;
use tauri::State;

/// Increment the counter `name` (e.g. `nav.prev`). Fire-and-forget; invalid
/// names are dropped silently.
#[tauri::command]
#[specta::specta]
pub fn telemetry_track(state: State<'_, AppState>, name: String) {
    crate::telemetry::record(&state, &name);
}

/// Set a gauge (current state, not a count), e.g. `lang = "da"`.
#[tauri::command]
#[specta::specta]
pub fn telemetry_set_gauge(state: State<'_, AppState>, name: String, value: String) {
    crate::telemetry::set_gauge(&state, &name, &value);
}

File: src/lib/telemetry.ts
import { commands } from "../bindings";

/**
 * Fire-and-forget usage counters for UI-only interactions (navigation clicks,
 * view toggles, export choices). Rust-side actions count themselves.
 *
 * Names are dotted lowercase identifiers (`nav.prev`, `export.format.json`);
 * anything else is dropped by the backend. Never pass user data as a name.
 * Counters are summed locally and sent once a day — see `src-tauri/src/telemetry.rs`
 * and the Telemetry section of the README for exactly what leaves the machine.
 */
export function track(name: string): void {
  commands.telemetryTrack(name).catch(() => {
    // Telemetry must never surface to the user.
  });
}

/** Record current state (not a count), e.g. `setGauge("lang", "da")`. */
export function setGauge(name: string, value: string): void {
  commands.telemetrySetGauge(name, value).catch(() => {
    // Telemetry must never surface to the user.
  });
}
